### DISCUSSION TITLE
[@types/lodash] `_.uniqBy` / type `[keyof T, any]` in `IterateeShorthand<T>` is not correct

### DISCUSSION THREAD

**Discussion by fix777** [DefinitelyTyped/DefinitelyTyped#64944] [Issues with a @types package]
If we use `x.y` as a value iteratee shorthand, the result is correct, but TS will complain.
```ts
uniqBy([{ x: { y: 1 } }, { x: { y: 2 } }, { x: { y: 2 } }], "x.y"); // the result is `[ { x: { y: 1 } }, { x: { y: 2 } } ]`
// Argument of type '"x.y"' is not assignable to parameter of type 'ValueIteratee<{ x: { y: number; }; }>'.typescript(2345)
```

If we use `['x', 'y']`, then TS will not complain, but the result is incorrect.
```ts
uniqBy([{ x: { y: 1 } }, { x: { y: 2 } }, { x: { y: 2 } }], ["x", "y"]); // the result is `[ { x: { y: 1 } } ]`
```

So my question is, is it possible to implement a custom value iteratee shorthand type for `_.uniqBy`?

Here is the reproduction [codesandbox](https://codesandbox.io/p/sandbox/quizzical-smoke-2x83no?selection=%5B%7B%22endColumn%22%3A74%2C%22endLineNumber%22%3A6%2C%22startColumn%22%3A3%2C%22startLineNumber%22%3A6%7D%5D&file=%2Fsrc%2Findex.ts).

Here are some screenshots just for references:
![image](https://user-images.githubusercontent.com/19484638/228782105-68382168-03d6-411b-b1fc-515ee70a20a0.png)
![image](https://user-images.githubusercontent.com/19484638/228779417-8464b5e8-5197-42c6-bca9-a0c604e83999.png)
![image](https://user-images.githubusercontent.com/19484638/228779687-04ea0a3d-6127-424b-a935-fc2305119dce.png)

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
Yeah looks like this type doesn't account for nested property names. We could change the key type to allow any string/number, but that would allow potential typos in property names to (incorrectly) compile successfully.

As a workaround, you can write you own function to get the nested property, which would be more type-safe anyway:

```ts
uniqBy([...], obj => obj.x.y);
```

Or if you really want to use the property shorthand (less type safe), you can explicitly call lodash's `property()` function:

```ts
uniqBy([...], property('x.y'));
```

**Reply by khitrenovich** [reply]
Yes, this is clearly about nested fields, here is [another reproduction](https://codesandbox.io/s/typescript-lodash-types-192-issue-ejw62i).

While I understand where the change comes from, I suspect it will break a lot of code (as it does for us), and both workarounds suggested 
 here go against the main principle of Lodash - code simplicity.

**Comment by bmarcotte** [comment]
I think we encountered the same, or a similar issue, but with groupBy().  We started getting type errors in version 4.14.192 specifically, with code that was working fine previously under 4.14.191.  Looks like it is likely due to the changes from PR #64722, and a rollback or fix is being discussed in PR #64922.  Hopefully, that is released sometime soon.

**Reply by khitrenovich** [reply]
Thank you for bringing this in, definitely the same root cause!
